### DISCUSSION TITLE
A `$` (dollar symbol) is being appended at the bottom of every page

### DISCUSSION THREAD

**Discussion by Jiralite** [remix-run/remix#10321] [Q&A]
When I turn on Cloudflare Web Analytics, Cloudflare injects a script which I suppose the following function is not expecting, because when it executes, a `$` is appended to my body on every page:

```JavaScript
function $RC(a, b) {
    a = document.getElementById(a);
    b = document.getElementById(b);
    b.parentNode.removeChild(b);
    if (a) {
        a = a.previousSibling;
        var f = a.parentNode
            , c = a.nextSibling
            , e = 0;
        do {
            if (c && 8 === c.nodeType) {
                var d = c.data;
                if ("/$" === d)
                    if (0 === e)
                        break;
                    else
                        e--;
                else
                    "$" !== d && "$?" !== d && "$!" !== d || e++
            }
            d = c.nextSibling;
            f.removeChild(c);
            c = d
        } while (c);
        for (; b.firstChild; )
            f.insertBefore(b.firstChild, c);
        a.data = "$";
        a._reactRetry && a._reactRetry()
    }
}
;$RC("B:0", "S:0")
```

Here's a reference to the start of the page load:

<img width="468" alt="image" src="https://github.com/user-attachments/assets/e4eae029-8d7d-4af3-879c-b925d353624f">

Cloudflare's script is injected above the `<template>`. This has the id `B:0`, which means this is what `$RC()` will immediately look for (see its first argument).

Whilst this function is being executed, it eventually reaches the final `for(; b.firstChild; )` loop. `a` is some text (`"\n"`), and the rest do not matter because `a` does not change. Once the loop breaks, it goes into `a`. `a` is just some text whose parent element is `<body>`. Now, it just does `a.data = "$"` and a dollar symbol is now at the bottom of the body on every page.

The issue here is that `a` for some reason is just some text. It seems `a` should instead be a comment node, as it looks like some hydration comment shenanigans are taking place.

Using the following dependencies:

- @remix-run/node: ^2.15.0
- @remix-run/react: ^2.15.0
- @remix-run/serve: ^2.15.0
- react ^18.2.0
- react-dom ^18.20

Needless to say, this will not occur during development. This problem is alleviated with Web Analytics turned off, but I'd rather have it on. So, whose issue is this? React's? Remix's? Not sure where to go from here.

**Comment by Hazqeel09** [comment]
facing the same issue, do you got any workaround for this or still the same?

**Reply by Jiralite** [reply]
I have found no workaround, sadly. I still also do not know who this should be reported to.

**Comment by bkyerv** [comment]
what is interesting is that it appears when the app is deployed on the vps. locally I don't see that $ sign rendered. I ensured that the env is production both locally and on the vps and yes locally I don't see it rendered

**Comment by Hazqeel09** [comment]
may i know the cpu architecture for the vps that you guys are using?  i am using arm but it didnt happen when i am deploying on my another vps using x86 architecture

**Comment by bkyerv** [comment]
~# uname -m
x86_64


I am also using cloudflare for domain set up so as OP said this most likely have to do with cloudflare doing something because locally I don't see that $ sign. mind that the app is dockerized

**Reply by Hazqeel09** [reply]
my both setup is quite similar, both is dockerized but my x86 didnt experience this but yuor answer rule out the cpu architecture

**Comment by Hazqeel09** [comment]
this is my current workaround, changed defer from cloudflare script to async, it work and i still got the web analytics eventho it does not show the web analytics are on inside the cloudflare

https://github.com/Hazqeel09/getdoav3/blob/main/remix/app/root.tsx

**Reply by bkyerv** [reply]
I see. thank you. in the code you shared there is a token - is it safe? did you mean to put that out there?

**Reply by adnanjaw** [reply]
Is that a realistic fix? I think Cloudflare should inject the script itself.

**Comment by gitnik** [comment]
any update on this?

**Comment by samohovets** [comment]
I can confirm that disabling Cloudflare Analytics like [this](https://developers.cloudflare.com/speed/speed-test/faq/#how-do-i-disable-real-user-monitoring-rum-if-it-has-been-enabled-from-the-observatory-test-result-page) resolves the issue.

But **instead of deleting it** completely, you can inject the JS script yourself.

**Comment by aschenkuttel** [comment]
Any update?

**Comment by yoni-noma** [comment]
Still happens :(

**Comment by accountunavaible** [comment]
disable cloudflare web analtics will resolve this issue

**Comment by gitmazya** [comment]
Same, had it in every app deployed on cloudflare, disabling Cloudflare Web Analytics helped

**Reply by Francismb** [reply]
This worked for me! Very odd.

**Comment by AdityaPrakash-26** [comment]
Happens for me too, and I don't have cloudflare analytics enabled... Any other work around?

**Reply by GabeMeister** [reply]
It eventually worked for me, you just have to wait a minute. @AdityaPrakash-26

**Comment by rossipedia** [comment]
From what I understand from reading [the docs](https://developers.cloudflare.com/pages/how-to/web-analytics/#enable-on-pages-project), CloudFlare automatically injects this script tag into your application's HTML output?

That's gonna break _any_ hydrated React Full Stack application, not just Remix/React Router.

@Hazqeel09 has the right approach here. Your app needs to be the source of truth for _all_ the HTML that is sent to the browser, so you should manage that `<script>` tag yourself. 

React's streaming SSR has certain requirements, one of them being that the arrangement of its control nodes (`<template>`, `<script>` and the `<!-- $ -->/<!-- /$ -->` comment markers) is _left alone_. CF injecting a script tag for you in the wrong place will cause bugs in React streaming SSR, which is what we're seeing here.

**Comment by GabeMeister** [comment]
This is also happening to our site https://www.redballoon.work/

**Comment by KristofaJosh** [comment]
Also happening on our site https://finnbud.com. 

I want the analytics on, so i guess the best fix is to add it explicitly and disable the automatic injection?
